### PR TITLE
[processor/k8sattribute] add more test cases for k8sattribute e2e

### PR DESCRIPTION
**Pull request by fatsheep9146**
Signed-off-by: Ziqi Zhao <zhaoziqi9146@gmail.com>

related issue #15651

### CODE REVIEW

File: processor/k8sattributesprocessor/e2e_test.go [hunk 1]
```diff
@@ -27,45 +28,99 @@ import (
...
+				}
+			}
+			return true
+		},
+	)
+
+	has := true
+	for _, v := range exists {
+		if !v {
+			has = false
```
Comment (frzifus) [inline processor/k8sattributesprocessor/e2e_test.go:121]:
```suggestion
			return false
```
nit, should be the same ^^

File: processor/k8sattributesprocessor/e2e_test.go [hunk 2]
```diff
@@ -27,45 +28,97 @@ import (
...
+			attrs: map[string]string{
+				"k8s.namespace.name": "default",
+				"k8s.job.name":       "test-telemetrygen-job",
+			},
+		},
+		{
+			name: "statefulset",
+			attrs: map[string]string{
+				"k8s.namespace.name":   "default",
+				"k8s.statefulset.name": "test-telemetrygen-statefulset",
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/e2e_test.go:48]: Can please you cover all attributes that are expected to be set by the k8sattributes processor by default? `k8s.pod.name` must be predictable for the statefulset. If we cannot predict some attribute values, can we check their presence?

**Review by dmitryax — COMMENTED** [review]
Nice addition! Thanks

File: processor/k8sattributesprocessor/e2e_test.go (hunk 2 above)
Comment (fatsheep9146) [reply processor/k8sattributesprocessor/e2e_test.go:48]: I enriched the compare mode into "equal", "regex" and "exist".

- equal for exactly matching
- regex for matching by regex pattern
- exist for checking if the key exists

Comment (dmitryax) [reply processor/k8sattributesprocessor/e2e_test.go]: Thank you @fatsheep9146 . It's super comprehensive coverage!

File: processor/k8sattributesprocessor/e2e_test.go [hunk 3]
```diff
@@ -19,53 +19,195 @@ package k8sattributesprocessor
...
+			return true
+		},
+	)
 
-	if err := scanner.Err(); err != nil {
-		assert.NoError(t, err)
+	for _, v := range exists {
+		if !v {
+			return false
+		}
```
[the remark this answers is not in the document]
Comment (fatsheep9146) [reply processor/k8sattributesprocessor/e2e_test.go:210]: now the failed message will be like

```
    e2e_test.go:181:
        	Error Trace:	/Users/zhaoziqi/go/src/go.opentelemetry.io/opentelemetry-collector-contrib/processor/k8sattributesprocessor/e2e_test.go:181
        	            				/Users/zhaoziqi/go/src/go.opentelemetry.io/opentelemetry-collector-contrib/processor/k8sattributesprocessor/e2e_test.go:142
        	Error:      	Received unexpected error:
        	            	k8s.labels.app attribute not found
        	Test:       	TestTraceE2E/daemonset
```

File: processor/k8sattributesprocessor/e2e_test.go [hunk 4]
```diff
@@ -12,60 +12,207 @@
...
-			)
+			service, exist := resourceSpans.Resource().Attributes().Get("service.name")
+			assert.Equal(t, true, exist, "span do not has 'service.name' attribute in resource")
+
+			if service.AsString() != expectedService {
+				continue
+			}
+
+			if err = resourceHasAttributes(resourceSpans.Resource(), kvs); err == nil {
+				return
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/e2e_test.go:177]: If we exit with nil, I think we should exit for any err as well, otherwise, we will be flooding the error output with duplicate messages

File: processor/k8sattributesprocessor/e2e_test.go [hunk 5]
```diff
@@ -12,60 +12,207 @@
...
+	)
 
-	if err := scanner.Err(); err != nil {
-		assert.NoError(t, err)
+	for k, v := range exists {
+		if !v {
+			return fmt.Errorf("%v attribute not found", k)
+		}
 	}
+	return nil
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/e2e_test.go:217]: Would be nice to accumulate all the not found attributes in one error. You can do

```suggestion
	var err error
	for k, v := range exists {
		if !v {
			err = multierr.Append(err, fmt.Errorf("%v attribute not found", k))
		}
	}
	return err
```

File: processor/k8sattributesprocessor/e2e_test.go [hunk 6]
```diff
@@ -12,60 +12,207 @@
...
+			if err = resourceHasAttributes(resourceSpans.Resource(), kvs); err == nil {
+				return
+			}
 		}
 	}
+	assert.NoError(t, err)
+}
+
+func resourceHasAttributes(resource pcommon.Resource, kvs map[string]*expectedValue) error {
+	exists := make(map[string]bool)
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/e2e_test.go:185]: `exists` is a bit confusing name. Maybe `foundAttrs`?

File: processor/k8sattributesprocessor/e2e_test.go [hunk 7]
```diff
@@ -12,60 +12,206 @@
...
 	unmarshaler := ptrace.JSONUnmarshaler{}
-	jobKeyFound := false
-	nsKeyFound := false
 	for scanner.Scan() {
-		traces, err := unmarshaler.UnmarshalTraces(scanner.Bytes())
-		assert.NoError(t, err)
+		traces, serr := unmarshaler.UnmarshalTraces(scanner.Bytes())
+		if serr != nil {
+			continue
+		}
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/e2e_test.go:163]: Why not `require.NoError(t, err)`? Is this common that unmarshaling can fail? I'm worried that the test suit can succeed if everything is skipped here

**Review by dmitryax — COMMENTED** [review]
Just one last comment

File: processor/k8sattributesprocessor/e2e_test.go (hunk 7 above)
Comment (fatsheep9146) [reply processor/k8sattributesprocessor/e2e_test.go:163]: Yes, I found that the last line of output traces something may be trimmed, which will cause this line is unable to unmarshalled correctly. I still do not know why.

But this only happens in last line. Other valid trace datas are enough to be tested.

File: processor/k8sattributesprocessor/e2e_test.go [hunk 8]
```diff
@@ -12,60 +12,206 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-//go:build e2e
-// +build e2e
-
```
Comment (dmitryax) [inline processor/k8sattributesprocessor/e2e_test.go:17]: Why is this removed?

File: processor/k8sattributesprocessor/e2e_test.go (hunk 7 above)
Comment (dmitryax) [reply processor/k8sattributesprocessor/e2e_test.go:163]: Ok, sounds good. let keep it like this for now

File: processor/k8sattributesprocessor/e2e_test.go (hunk 8 above)
Comment (fatsheep9146) [reply processor/k8sattributesprocessor/e2e_test.go:17]: > Why is this removed?

My bad, I test it locally

Comment (fatsheep9146) [reply processor/k8sattributesprocessor/e2e_test.go]: add it backed, let's wait for all checks are passed.

**Review by dmitryax — APPROVED** [review]
LGTM
